checker/src/tag_domain.rs

```diff
@@ -8,92 +8,48 @@ use crate::bool_domain::BoolDomain;
 use log_derive::logfn_inputs;
 use mirai_annotations::*;
 use rpds::{rbt_map, RedBlackTreeMap};
-use rustc_hir::def_id::{CrateNum, DefId, DefIndex};
-use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
-
-/// A replication of the `DefId` type from rustc. The type is used to implement serialization.
-#[derive(Ord, PartialOrd, Eq, PartialEq, Copy, Clone, Hash)]
-pub struct SerializableDefId {
-    pub krate: CrateNum,
-    pub index: DefIndex,
-}
-
-impl From<DefId> for SerializableDefId {
-    fn from(did: DefId) -> SerializableDefId {
-        SerializableDefId {
-            krate: did.krate,
-            index: did.index,
-        }
-    }
-}
-
-impl std::fmt::Debug for SerializableDefId {
-    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.write_fmt(format_args!(
-            "{:?}",
-            DefId {
-                krate: self.krate,
-                index: self.index
-            }
-        ))
-    }
+use serde::{Deserialize, Serialize};
+use std::rc::*;
+
+/// Tags are either a pre-defined tag for constant-time security, or user-defined ones.
+#[derive(Ord, PartialOrd, Eq, PartialEq, Debug, Clone, Serialize, Deserialize, Hash)]
+pub enum Tag {
+    /// A tag for analyzing constant-time security. The type is pre-defined in mirai_annotations.
+    ConstantTime,
+
+    /// A user-defined tag is represented as a pair of its tag kind and its propagation set.
+    /// The tag kind is the path of the declared tag type, and the propagation set is associated to the
+    /// tag type as a const generic parameter.
+    Custom {
+        def_path_str: Rc<String>,
```

Comment:
You should rather use the KnownNames enum here.

Reply:
The tag types are defined by users. I used to use DefId here.